common/network-common/src/main/java/org/apache/spark/network/crypto/GcmTransportCipher.java

```diff
@@ -0,0 +1,356 @@
+
+        @Override
+        public long position() {
+            return 0;
+        }
+
+        @Override
+        public long transferred() {
+            return transferred;
+        }
+
+        @Override
+        public long count() {
+            return encryptedCount;
+        }
+
+        @Override
+        public long transferTo(WritableByteChannel target, long position) throws IOException {
+            Preconditions.checkArgument(position == transferred(),
+                    "Invalid position.");
+            int transferredThisCall = 0;
+            // The format of the output is:
+            // [8 byte length][Internal IV and header][Ciphertext][Auth Tag]
+            if (!headerWritten) {
+                ByteBuffer expectedLength = ByteBuffer
+                        .allocate(LENGTH_HEADER_BYTES)
+                        .putLong(encryptedCount)
+                        .flip();
+                target.write(expectedLength);
+                int headerWritten = LENGTH_HEADER_BYTES + target.write(encrypter.getHeader());
+                transferredThisCall += headerWritten;
+                this.transferred += headerWritten;
+                this.headerWritten = true;
+            }
+
+            while (bytesRead < bytesToRead) {
+                long readableBytes = getReadableBytes();
+                boolean lastSegment = readableBytes <= plaintextBuffer.capacity();
+                plaintextBuffer.clear();
+                int readLimit =
+                        (int) Math.min(readableBytes, plaintextBuffer.capacity());
+                plaintextBuffer.limit(readLimit);
+                if (plaintextMessage instanceof ByteBuf byteBuf) {
+                    byteBuf.readBytes(plaintextBuffer);
+                    long inputBytesRead = readableBytes - byteBuf.readableBytes();
+                    bytesRead += inputBytesRead;
+                } else if (plaintextMessage instanceof AbstractFileRegion fileRegion) {
+                    ByteBufferWriteableChannel plaintextChannel =
+                            new ByteBufferWriteableChannel(plaintextBuffer);
+                    long transferred =
+                            fileRegion.transferTo(plaintextChannel, fileRegion.transferred());
+                    bytesRead += transferred;
+                    if (transferred == 0) {
+                        // File regions may return 0 if they are not ready to transfer
+                        // more data. In that case, we'll return with the expectation
+                        // that this transferTo() is called again.
+                        return transferredThisCall;
+                    }
+                }
+                plaintextBuffer.flip();
+                ciphertextBuffer.clear();
+                try {
+                    encrypter.encryptSegment(plaintextBuffer, lastSegment, ciphertextBuffer);
+                } catch (GeneralSecurityException e) {
+                    throw new RuntimeException(e);
```

Comment:
There are a few Java exceptions that are thrown by the underlying crypto: `IllegalKeyExceptions` and `ShortBufferExceptions` for example.

In practice, I don't see a way these would get thrown. We're generating the key and buffers ourselves. So, if an exception is thrown it is truly unexpected and not something we can recover from.

`IllegalStateException` would be okay because somehow we would have ended up with a corrupted key or buffer or something.